### PR TITLE
[WIP] feat(html): add HTML module support (`html`)

### PR DESCRIPTION
**Pull request by michael-ciniawsky**
This PR attempts to add support for a new module type (`module.type === 'html'`)  to support HTML 'natively', replacing the need for loaders/plugins like the `html-loader/html-webpack-plugin` and  **adding support for using HTML as an Entrypoint** and to declare Entrypoints  (JS, CSS, HTML) from within the HTML Template

**`entry.html`**
```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>HTML Modules</title>
    <!-- Entrypoint Support (CSS) -->
    <link href="./entry.css" rel="stylesheet" entry>
  </head>
  <body>
    <div>
      <!-- HTML Module Support -->
      <import src="./import.html"></import>

      <!-- Asset Module Support -->
      <img src="./assets/file.png">

      <!-- Entrypoint Support (HTML) -->
      <a href="./entry.html" entry></a>

      <!-- Entrypoint Support (JS) -->
      <script src="./entry.js" entry></script>
    </div>
  </body>
</html>
```

**`webpack.config.js`**
```js
const config = {
  entry: {
    index: 'path/to/entry.html'
  },
  module: {
    rules: [
       {
           test: /\.html$/,
           type: 'html'
       },
       ...
    ]
  }
}
```

> :information_source: This PR includes defaults for common `module.rules`, so normally it won't be necessary to specify `module.rules` for common cases (`.html`) like in the example above, it's sole purpose is to illustrate the concept, adding custom rules for e.g other templating languages (`.pug` etc) is of course supported aswell

---
### Fix

- [ ] Render HTML correctly 
   - `HTMLTemplate`
   - `HTMLGenerator` (Dependencies)
- [ ] Replace entrypoint placeholders with `output` URLs (Assets URLs)
   - needs to rerender on every change, alonside the changes files

### Feature

- [ ] Figure out how to treat/handle a HTML `ChunkGroup/Chunk` (Entrypoint)
- [ ] `output` (`webpack.config.js`) Support (e.g `[name].[ext]`)
   - no `[hash]` placeholder support (doesn't make sense)
   - the 'main'  entry should ideally (always) be named `index` (`index.html`)
- [ ] `preload`  Support
- [ ] `prefetch` Support
  - `<a rel="next" href="./entry.html" entry>` (maybe)

### Chore

- [ ] Add Type Definitions

### Test

- [ ] Add tests
---

### CODE REVIEW

File: lib/WebpackOptionsDefaulter.js [hunk 1]
```diff
@@ -56,6 +56,10 @@ class WebpackOptionsDefaulter extends OptionsDefaulter {
 			{
 				test: /\.wasm$/i,
 				type: "webassembly/experimental"
+			},
+			{
+				test: /\.html$/,
```
Comment (kingdaro) [inline lib/WebpackOptionsDefaulter.js]: The other RegExps have the `i` flag but this one doesn't. Oversight?

Comment (michael-ciniawsky) [reply lib/WebpackOptionsDefaulter.js]: Likely :)

File: test/configCases/html/module/index.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import html from './index.html';
+
+it('Should import HTML', function() {
+	html.should.be.eql(`<div id="app"></div>`);
+});
+
+it('Should import() HTML', function() {
+	import('./index.html').then((html) => {
```
Comment (Janpot) [inline test/configCases/html/module/index.js]: I think you need to return this promise from the test.

File: lib/Template.js [hunk 3]
```diff
@@ -202,4 +202,43 @@ module.exports = class Template {
 		}
 		return source;
 	}
+
+	static renderHTMLChunk(chunk, filterFn, moduleTemplate, dependencyTemplates) {
```
Comment (glen-84) [inline lib/Template.js]: Shouldn't this be `renderHtmlChunk`? (etc.)

It's `stats.toJson()`, `JsonGenerator`, `JsonParser`, etc., so acronyms are written in sentence case.

**Comment by glen-84** [comment]
Will this PR resolve #536?

I'm new to webpack, but I'd like to be able to point it to a list of HTML files, and have each script/image/stylesheet in those files processed "automatically".

For example, given:

```js
entry: "page1.html"
```

```html
<!DOCTYPE html>
<html>
<head>
    <title>Example</title>
    <script src="main.ts"></script>
</head>
<body>
    <img src="image.png" />
</body>
</html>
```

I'd like `main.ts` to be processed as an entry file, and for its output path to be written into the resulting HTML (by default), like:

```html
...
<script src="/dest/main.js"></script>
...
```

The image would also be processed in a similar way (as well as any link/stylesheet elements, etc.).

(In my specific case, I'd probably have to block the output of the HTML itself, because I'm working with server-side HTML templates, but I assume that I could do something like this with a plugin?)

I've just started work on implementing webpack, so I'd like to know if this type of thing will be supported "soon", or if I should just use the current method of importing CSS inside JS, etc. (which I find a bit strange).

File: lib/WebpackOptionsDefaulter.js [hunk 4]
```diff
@@ -113,6 +117,8 @@ class WebpackOptionsDefaulter extends OptionsDefaulter {
 			}
 			return "[id].js";
 		});
+		// HTML
+		this.set("output.htmlModuleFilename", "[modulehash].module.html");
```
Comment (michael-ciniawsky) [inline lib/WebpackOptionsDefaulter.js]: remove

File: lib/WebpackOptionsDefaulter.js [hunk 5]
```diff
@@ -317,7 +323,7 @@ class WebpackOptionsDefaulter extends OptionsDefaulter {
 		this.set("resolve", "call", value => Object.assign({}, value));
 		this.set("resolve.unsafeCache", true);
 		this.set("resolve.modules", ["node_modules"]);
-		this.set("resolve.extensions", [".wasm", ".mjs", ".js", ".json"]);
+		this.set("resolve.extensions", [".html", ".wasm", ".mjs", ".js", ".json"]);
```
Comment (michael-ciniawsky) [inline lib/WebpackOptionsDefaulter.js]: remove

**Comment by glen-84** [comment]
It's now possible to vote for this feature @ https://webpack.js.org/vote (`HTML as entrypoint`).

Note: There is currently a bug with the voting app (https://github.com/webpack-contrib/voting-app/issues/15).

**Pull request closed without merging**